client/pd_service_discovery.go

```diff
@@ -106,61 +105,65 @@ type pdBaseClient struct {
 	clusterID uint64
 	// addr -> a gRPC connection
 	clientConns sync.Map // Store as map[string]*grpc.ClientConn
-	// dc-location -> TSO allocator leader URL
-	tsoAllocators sync.Map // Store as map[string]string
 
 	// leaderSwitchedCallbacks will be called after the leader swichted
 	leaderSwitchedCallbacks []func()
-	// membersChangedCallbacks will be called after there is any membership
-	// change in the leader and followers
+	// membersChangedCallbacks will be called after there is any membership change in the
+	// leader and followers
 	membersChangedCallbacks []func()
-	// tsoAllocatorLeaderSwitchedCallback will be called when any global/local
-	// tso allocator leader is switched.
-	tsoAllocatorLeaderSwitchedCallback []func()
+	// tsoAllocatorLeadersUpdatedCallbacks will be called when the global/local tso allocator
+	// leader list is updated. The input is a map {DC Localtion -> Leader Addr}
+	tsoAllocatorLeadersUpdatedCallbacks []tsoAllocServingAddrUpdatedFunc
 
 	checkMembershipCh chan struct{}
 
-	wg     *sync.WaitGroup
-	ctx    context.Context
-	cancel context.CancelFunc
+	wg        *sync.WaitGroup
+	ctx       context.Context
+	cancel    context.CancelFunc
+	closeOnce sync.Once
 
 	tlsCfg *tlsutil.TLSConfig
 	// Client option.
 	option *option
 }
 
-// newPDBaseClient returns a new baseClient.
-func newPDBaseClient(ctx context.Context, cancel context.CancelFunc,
-	wg *sync.WaitGroup, urls []string, tlsCfg *tlsutil.TLSConfig, option *option) BaseClient {
-	bc := &pdBaseClient{
+// newPDServiceDiscovery returns a new baseClient.
+func newPDServiceDiscovery(ctx context.Context, cancel context.CancelFunc,
+	wg *sync.WaitGroup, urls []string, tlsCfg *tlsutil.TLSConfig, option *option) *pdServiceDiscovery {
+	pdsd := &pdServiceDiscovery{
 		checkMembershipCh: make(chan struct{}, 1),
 		ctx:               ctx,
 		cancel:            cancel,
 		wg:                wg,
 		tlsCfg:            tlsCfg,
 		option:            option,
 	}
-	bc.urls.Store(urls)
-	return bc
+	pdsd.urls.Store(urls)
+	return pdsd
 }
 
-func (c *pdBaseClient) Init() error {
-	if err := c.initRetry(c.initClusterID); err != nil {
-		c.cancel()
-		return err
-	}
-	if err := c.initRetry(c.updateMember); err != nil {
-		c.cancel()
-		return err
+func (c *pdServiceDiscovery) Init() error {
+	if !c.isInitialized {
```

Comment:
What does the flag use for?

Reply:
microservice mode:
client.pdsd = newPDServiceDiscoverytsoClient.sd

Reply:
BTW, I didn't init and Close() both pdServiceDiscovery and tsoMcsDiscovery in the client, because client doesn't save the object of tsoMcsDiscovery.

Reply:
Now the user calls NewTSOClientWithContext to enter into tso microservice mode, after we enable client-side discovery for service configuration and state, this code should be get clearer.